client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -349,13 +351,16 @@ public void dataAppended(DataAppended dataAppended) {
             log.trace("Received dataAppended ack: {}", dataAppended);
             long ackLevel = dataAppended.getEventNumber();
             long previousAckLevel = dataAppended.getPreviousEventNumber();
-            try {
-                checkAckLevels(ackLevel, previousAckLevel);
-                state.noteSegmentLength(dataAppended.getCurrentSegmentWriteOffset());
-                ackUpTo(ackLevel);
-            } catch (Exception e) {
-                failConnection(e);
-            }
+            // invoke state updation on a different thread and release the EventLoop thread.
+            dataAppendedProcessor.add(() -> CompletableFuture.runAsync(() -> {
```

Comment:
I think we can actually simplify this a good deal, and avoid the future chaining.

If we combine: `state.removeInflightBelow` and `state.releaseIfEmptyInflight` into a single method, `ackUpTo` will  only need to call into the state object and aquire the lock once. Then  we can do a `runAsync` to looping over the acks and completing their futures. This means we don't need the SequentialProcessor at all, so we don't need to construct chains of futures.

Reply:
Thanks, @tkaitchuck. I have updated the `ackUpTo` to ensure it needs to acquire the lock only once. (i.e. `state.removeInflightBelow` and `state.releaseIfEmptyInflight` are invoked under the same re-entrant lock).

On receiving a DataAppended the writer performs these 3 operations.
- `checkAckLevels(ackLevel, previousAckLevel);`
- `state.noteSegmentLength(dataAppended.getCurrentSegmentWriteOffset())`;
- `ackUpTo(ackLevel)`; // now optimized based on the comment.
All three actions need to be performed on a thread other than the `eventLoop` thread. Hence, I do not see a way to eliminate the `SequentialProcessor` which creates chain of futures. Let me know if I have misunderstood your comment...

Reply:
I think you missed my suggestion. I was suggesting the ackUpTo which currently does this:
```
for (PendingEvent toAck : state.removeInflightBelow(ackLevel)) {
  if (toAck.getAckFuture() != null) {
   toAck.getAckFuture().complete(null);
  }
  toAck.getData().release();
}
```
could be doing this
```
var toAck = state.removeInflightBelow(ackLevel);
// The above needs the lock and to be ordered.
// The below does not need a lock and does not need to be ordered.
for (PendingEvent toAck : toAck) {
  if (toAck.getAckFuture() != null) {
   toAck.getAckFuture().complete(null);
  }
  toAck.getData().release();
```
If we move that `for` loop out to a background thread, then the rest of this code can happen in the event loop thread.

Reply:
updated the code, now the future completion and releasing the ByteBuf does not happen on the eventloop thread.